### PR TITLE
Adjust GitHub and Jira extraction to a consistent format

### PR DESCRIPTION
**Pull request by fehnkera**
'issue_processing.py' and 'jira_issue_processing.py' get adjusted, so
they both can print a '.list' file with a consistent format. The new
format is:
"issue number";"issue name";"[issue types]";"issue state";"issue creation
date";"issue resolve date";"[issue components]";"event type";"event user
name";"event user e-mail";"event time";"event info 1";"event info 2"

Attached diagram gives a overview about how the extraction to the new format works:
[codeface-data-extraction-diagramm.pdf](https://github.com/se-passau/codeface-extraction/files/2265443/codeface-data-extraction-diagramm.pdf)

For now only the methods 'print_to_disk_new' ('issue_processing.py') and
'print_to_disk_bugs' ('jira_issue_processing.py') use this new format,
because the network library only supports the old format. When the
network library is updated all output methods shall use this format.

This fixes #91.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

### CODE REVIEW

**Review by ecklbarb — CHANGES_REQUESTED** [review]
Thank you for this nice additions,  @fehnkera!

File: issue_processing/issue_processing.py
Comment (ecklbarb) [inline issue_processing/issue_processing.py]: Please use consistent quotation marks.

File: issue_processing/issue_processing.py
Comment (ecklbarb) [inline issue_processing/issue_processing.py]: See the comment above.

File: issue_processing/issue_processing.py
Comment (ecklbarb) [inline issue_processing/issue_processing.py]: Can you check as first step, whether the date has the correct format?

File: issue_processing/issue_processing.py
Comment (ecklbarb) [inline issue_processing/issue_processing.py]: Maybe, you can find an alternative to the try-except.

File: issue_processing/issue_processing.py
Comment (ecklbarb) [inline issue_processing/issue_processing.py]: I think, the code would be clearer, if you extract the parsing of the event information in a separate method.

File: issue_processing/issue_processing.py
Comment (ecklbarb) [inline issue_processing/issue_processing.py]: Please use empty spaces before and after `=`.

File: issue_processing/issue_processing.py
Comment (ecklbarb) [inline issue_processing/issue_processing.py]: Please, use empty spaces before and after `=` and consistent quotation marks.

File: issue_processing/issue_processing.py
Comment (ecklbarb) [inline issue_processing/issue_processing.py]: Please, make a new line after `"""`.

File: issue_processing/issue_processing.py
Comment (ecklbarb) [inline issue_processing/issue_processing.py]: Good hint. Thank you for this comment.

File: issue_processing/issue_processing.py
Comment (ecklbarb) [inline issue_processing/issue_processing.py]: ` """`
`Print issues to file 'issues_new.list' in result folder.`

File: issue_processing/jira_issue_processing.py
Comment (ecklbarb) [inline issue_processing/jira_issue_processing.py]: Please, use consistent quotation marks.

**Review by ecklbarb — COMMENTED** [review]
I commented on many small stuff. Basically, your changes look really fine.

File: issue_processing/issue_processing.py
Comment (clhunsen) [reply issue_processing/issue_processing.py]: Before I look at the details of this PR myself, one small comment and advice here: Python has a style guide ([PEP8](https://www.python.org/dev/peps/pep-0008/); unlike R), which is also directly supported by PyCharm. You just need to re-format the file (press <kbd>Alt+Shift+L</kbd>).
I highly recommend using PyCharm's formatter. This will solve most of the formatting issues up-front (except for quotation marks, I suppose).

Why I comment here: The style guide states the following – in contrast to our coding conventions from the network library:
> Don't use spaces around the = sign when used to indicate a keyword argument or a default parameter value.

File: issue_processing/issue_processing.py
Comment (clhunsen) [reply issue_processing/issue_processing.py]: Instead of the `try-except` cascade, we may try something like [this](https://stackoverflow.com/a/14245142):
```python
from dateutil import parser as dateparser
dateparser.parse(string)
```

Otherwise, there is a suggestion in [another answer of the same question](https://stackoverflow.com/a/14245134) on how to use a loop instead of the current `try-except` cascade (maybe with some more documentation :wink:):
```python
for date_format in DATE_FORMATS:
    try:
        my_date = datetime.strptime(test_date, date_format)
    except ValueError:
        pass
    else:
      break
else:
  my_date = None
```

File: issue_processing/issue_processing.py
Comment (clhunsen) [inline issue_processing/issue_processing.py]: The line breaks here look weird. If you have some grouping here, please insert appropriate comments; otherwise, just re-flow the line breaks.

File: issue_processing/issue_processing.py
Comment (clhunsen) [reply issue_processing/issue_processing.py]: Maybe, we should do that also for the information on related issues and related commits.

File: issue_processing/issue_processing.py
Comment (clhunsen) [inline issue_processing/issue_processing.py]: its

File: issue_processing/issue_processing.py
Comment (clhunsen) [inline issue_processing/issue_processing.py]: `known_resolutions` and also `known_types` are predestined to be global variables. Please move them outside of this function.

File: issue_processing/issue_processing.py
Comment (clhunsen) [inline issue_processing/issue_processing.py]: In the last data file I sent you, there is an event called `unlabeled`.

File: issue_processing/issue_processing.py
Comment (clhunsen) [inline issue_processing/issue_processing.py]: Please reference the function here that writes this particular file.

File: issue_processing/issue_processing.py
Comment (clhunsen) [reply issue_processing/issue_processing.py]: Please add a `TODO` in front of this line, so we can easily re-find this comment.

File: issue_processing/issue_processing.py
Comment (clhunsen) [inline issue_processing/issue_processing.py]: Please add a `TODO` in front of this line, so we can easily re-find this comment.

File: issue_processing/issue_processing.py
Comment (clhunsen) [inline issue_processing/issue_processing.py]: `issues_new.list` :wink:

File: issue_processing/jira_issue_processing.py
Comment (clhunsen) [inline issue_processing/jira_issue_processing.py]: Please move this function to the global level as it is a utility function.

File: issue_processing/jira_issue_processing.py
Comment (clhunsen) [inline issue_processing/jira_issue_processing.py]: What is "the new GitHub format"? :wink:  Please refer to the corresponding function in the other file.

**Review by clhunsen — CHANGES_REQUESTED** [review]
Thank you very much for your work, @fehnkera! Overall, this looks terrific. :grinning: 

Unfortunately, there are some things to improve before we can merge (rather structural stuff, no  functional issues):
- Consider [this comment](https://github.com/se-passau/codeface-extraction/pull/16/files#r210663617) and reformat all changed files.
- Work on the other comments by Barbara and me.
- Rebase your commits and rewrite the commit messages for better structure (especially the commit titles, see [here](https://github.com/se-passau/codeface-extraction-r/blob/8c2a8255966cc6c38ea16e64a28b135ef8456e58/CONTRIBUTING.md#commit-messages) for recommendations). Using a more consistent formatting really helps with debugging in the future and also helps yourself to structure your changes and thoughts.

    For example, you can easily use an itemization in commit 9602adbec8012abfb54b6a2d0dedcce855e29805. This way, the columns in the new format can be parsed better by the reader.
- Please add the PDF file (and also the GRAPHML source file) for the diagram to the repository.

If you have any questions or need help, please do not hesitate to ask!

File: issue_processing/issue_processing.py
Comment (ecklbarb) [reply issue_processing/issue_processing.py]: Thank you for this hint @clhunsen.

**Comment by ecklbarb** [comment]
Regarding the last point to change, @bockthom and I asked @fehnkera not to add the diagram to the repository, but to include it in the pull request.

**Comment by clhunsen** [comment]
> Regarding the last point to change, @bockthom and I asked @fehnkera not to add the diagram to the repository, but to include it in the pull request.

Ah, I have not known about that. Okay, then forget about the last point in my review, @fehnkera. Maybe, we find another (local) place (in the NFS) where to put it for later use.

File: issue_processing/issue_processing.py
Comment (fehnkera) [reply issue_processing/issue_processing.py]: The dateutil parser works. Thank you very much for the suggestion @clhunsen

File: issue_processing/issue_processing.py
Comment (bockthom) [inline issue_processing/issue_processing.py]: Is there a reason for the plural in `empty Lists`? If not, could you please use the singular here?
In addition, adding a comma after the if-clause (i.e., after `eventsList`) will optimize the flow of reading.

This comment applies also for all the following comments regarding empty lists below.

File: issue_processing/issue_processing.py
Comment (bockthom) [inline issue_processing/issue_processing.py]: Meanwhile, the GH-Wrapper does provide the information about time and user of related issues (it is not merged to master yet, but we already provided some example data on the new format, see here: https://github.com/se-passau/GitHubWrapper/issues/22#issuecomment-411324641
So, please include this (have a look at the most recent example data we have sent you).

File: issue_processing/issue_processing.py
Comment (bockthom) [inline issue_processing/issue_processing.py]: Please differentiate between referenced commits and commits which are part of a pull request:
- "relatedCommits": have `Username` (and name, email) and `Referenced_at` in the GH
- "hasCommit": `Username` (and name, email) and `Referenced_at` are `null` in the GH data, leading to empty strings for `user` and `created_at` here.

File: issue_processing/issue_processing.py
Comment (bockthom) [inline issue_processing/issue_processing.py]: There should already be examples for `unlabeled` in the most recent GH data. If not, let us now.

File: issue_processing/jira_issue_processing.py
Comment (bockthom) [inline issue_processing/jira_issue_processing.py]: maybe also add a `TODO` here as the `new` parts should replace the current ones later, I guess. (Or I am wrong here?)

This applies also for the following comments below.

**Review by bockthom — CHANGES_REQUESTED** [review]
Thanks for all the nice changes, @fehnkera. I really like the new format very well!

I just have a few comments regarding your changes, please have a look at them. (Some comments address the new format regarding referencing issues, referenced commits, and "unlabeled" events. I am not sure if you are currently working on them or just have forgotten about them -- anyway, just take my remarks on them as a friendly reminder.)

In addition, something went wrong with the rebase of the commits: Now, there are some commits not authored by you appearing in this pull-request. Those commits are already part of the master branch and should not be added again. So, please rebase again on the current version of master and remove the duplicated commits introduced by you. If you need help with the rebase, no worries, just let us know.

**Comment by bockthom** [comment]
> > Regarding the last point to change, @bockthom and I asked @fehnkera not to add the diagram to the repository, but to include it in the pull request.
> 
> Ah, I have not known about that. Okay, then forget about the last point in my review, @fehnkera. Maybe, we find another (local) place (in the NFS) where to put it for later use.

Yes, I would prefer a local place. We can decide upon that when everything related to this PR is finished.

@fehnkera Could you please also update the diagram? If I remember correctly, the most recent changes are not yet updated in the diagram.

**Comment by clhunsen** [comment]
> In addition, something went wrong with the rebase of the commits: Now, there are some commits not authored by you appearing in this pull-request. Those commits are already part of the master branch and should not be added again. So, please rebase again on the current version of master and remove the duplicated commits introduced by you. If you need help with the rebase, no worries, just let us know.

On this, @fehnkera sent an e-mail already, I will take care of this.

File: issue_processing/jira_issue_processing.py
Comment (ecklbarb) [inline issue_processing/jira_issue_processing.py]:
```
"""
Parse issues from the xml-data
```
Maybe, you can add a line break here to get consistent comments. Thank you, @fehnkera !

File: issue_processing/jira_issue_processing.py
Comment (ecklbarb) [reply issue_processing/jira_issue_processing.py]: @clhunsen, what do you think: Which quotations marks should be use to be consistent?

File: issue_processing/jira_issue_processing.py
Comment (ecklbarb) [inline issue_processing/jira_issue_processing.py]: It is a good idea to add the function `create_user`. :)

**Review by ecklbarb — COMMENTED** [review]
Overall, its look really great. Thank you, @fehnkera!

File: issue_processing/jira_issue_processing.py
Comment (clhunsen) [reply issue_processing/jira_issue_processing.py]: Does not matter to me as long as they are consistent in a single file – it makes reading the source code easier. Personally, I prefer double quotation marks.

File: issue_processing/issue_processing.py
Comment (clhunsen) [inline issue_processing/issue_processing.py]: Is there a reason why this event is written in camelCase instead of snake_case?

**Review by clhunsen — APPROVED** [review]
Nice work, @fehnkera. This PR looks good to me, I just have one minor comment.

File: issue_processing/jira_issue_processing.py
Comment (fehnkera) [reply issue_processing/jira_issue_processing.py]: The problem with the not consistent quotation marks is in the whole file, not only the parts of the file that I worked on. Shall I also change the quotation marks in the other methods?

File: issue_processing/issue_processing.py
Comment (fehnkera) [reply issue_processing/issue_processing.py]: The camelCase name is the name @bockthom suggested. But I can change it to snake_case if preferred.

File: issue_processing/jira_issue_processing.py
Comment (clhunsen) [reply issue_processing/jira_issue_processing.py]: Yes, please. A quick search-replace call will probably suffice.

File: issue_processing/issue_processing.py
Comment (clhunsen) [reply issue_processing/issue_processing.py]: I was just wondering because all other event names I found have a snake_case pattern (at least, in the overview file linked in the PR's first post). I am fine with the name, but I suggest to convert it to the snake_case variant `has_commit` for consistency. :wink:

**Comment by fehnkera** [comment]
The updated diagram with the new events:

[codeface-data-extraction-diagramm.pdf](https://github.com/se-passau/codeface-extraction/files/2353126/codeface-data-extraction-diagramm.pdf)

**Comment by clhunsen** [comment]
Thanks for the tedious work on the last two commits and the updated diagram, @fehnkera. 

For me, this PR looks fine now. @ecklbarb, do you have anything else to discuss or improve? I will wait for your approval before merging.

**Comment by ecklbarb** [comment]
Thank you for the whole work you have done, @fehnkera.

For me, this PR looks fine, too. @clhunsen, you can merge the changes into the master.

**Review by ecklbarb — APPROVED** [review]

**Comment by bockthom** [comment]
@fehnkera Also a thank-you from mine. Good work! 
(Sorry for being inconsistent with my suggestion regarding "has_commits" using camel case, unfortunately I had not checked which coding style was used in the file before.)

Nevertheless, I have one question regarding the updated diagram:

> The updated diagram with the new events:
> [codeface-data-extraction-diagramm.pdf](https://github.com/se-passau/codeface-extraction/files/2353126/codeface-data-extraction-diagramm.pdf)

Does the diagram already comprise the new differentiation between referenced commits and the new "has_commits" event? At least, I cannot find this event in the updated diagram. Maybe it should not appear there and I am wrong or I have overlooked it, but if this event is missing there, please add it in the diagram! Thanks a lot!

**Comment by fehnkera** [comment]
Thank you for your note @bockthom. You are right, the differentiation between referenced commits and "has_commits" was missing in the diagram. So here is an updated version with this event in it:
[codeface-data-extraction-diagramm.pdf](https://github.com/se-passau/codeface-extraction/files/2375135/codeface-data-extraction-diagramm.pdf)

**Comment by bockthom** [comment]
Thanks for updating the diagram again. Looks good now!

**Pull request merged**
